_drafts/hw_actors.md

```diff
@@ -0,0 +1,150 @@
+
+<!-- excerpt start -->
+
+This article is intended to describe the alternative approach somewhere in between RTOS and bare-metal no-OS systems.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+
+## Hardware scheduler
+
+
+Do you remember how the 'usual' RTOS scheduler works? It maintains a set of executable entities with some user-defined priority. These entities may activate each other either directly or indirectly. The scheduler takes responsibility that the most prioritized entity will 'run' and activation of other entities will trigger 'rescheduling' and possibly preemption of the currently active one.
+
+If you look at the programmer's model of interrupt controllers like NVIC or GIC, you notice that these devices do exactly that. For example, in the NVIC case, the device contains a special register called STIR (Software Trigger Interrupt Register). Writing a vector number to that register causes corresponding interrupt to be activated exactly the same way as hardware does. Since the user may also adjust interrupt preemption priority, interrupt controller itself may be treated as implementation of the so-called [actor model](https://en.wikipedia.org/wiki/Actor_model) in **hardware**. When you trigger an interrupt its priority management (i.e. delaying when interrupt with greater priority is active) and activation is also performed by hardware. So, necessary ingredients for the hardware-implemented actor model are:
+
+- Ability to trigger interrupts with different vectors programmatically
+- Independently adjustable priority for each vector
+
+The approach is to use this model and map priorities to interrupt vectors with user-defined priority. Note that this model contains no threads and runs at a single shared stack. It is inherently asynchronous and even applicable to multiprocessor systems since (in the case of GIC) you may trigger interrupts on other CPUs too.
+
+
+## The concept
+
+
+Since the actor model itself is a very complex topic for a single pocket article, instead of talking about abstractions, I mostly focus on their representation in the MCU environment.
+
+In the classic actor model, an actor is just a message handler similar to interrupt handlers. Unlike threads, it has no dedicated stack and its purpose is to react to messages arriving into its internal mailbox.
```

Comment:
Yes, they all share the interrupt stack since they are essentially run-to-completion routines. I mean there they do not need separate stack as threads do.